refact(api): Make streams/sessions controllers typescript

victorges opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.com/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->
**What does this pull request do? Explain your changes. (required)**
This adds typing to a lot of stuff, using `/stream` API controller as starting point.

Starting point because, on the way to typing that controller as much as possible,
I also bumped into a couple helpers or dependencies that also weren't typed yet
so I needed to go deeper.

Also typed all of its tests, which was mostly some repetitive changes, but also  ended
up creating new branches of typing stuff like all the `index.ts` and `parse-cli.ts` so the
test server/clients were fully typed as well.

**Specific updates (required)**
 - Add type checking to build pipeline (including `tsconfig.json` for the package)
 - Migrate to typescript:
   - `controllers/stream.ts` and its tests
   - `controllers/session.ts`
   - `middleware/` `hardcocded-nodes` and `kubetrnetes`, and their respective tests
   - `parse-cli.ts` (I recommend reviewing this one from this commit: https://github.com/livepeer/livepeer-com/pull/476/commits/0297730f626b4f3b0221d65120dbff8656639696?w=1)
   - `util.ts` (which involved some hacky typing of isomorphic fetch, but I could migrate to node-fetch completely if preferred)
 - Also some typing and sometimes actual code fixes on the way
 
## -

- **How did you test each of these updates (required)**
`yarn && yarn test --silent`

I bet the actual code transpiled in `dist/` has changed very little. 
I could make that safety check as well, let me know.

**Does this pull request close any open issues?**
I don't think so. Just some housekeeping in the project.

**Checklist:**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

victorges committed:
Make streams controller TypeScript

Only missing some getIngest part which I
think will involve some new things on the
request, will be better to keep those separate.

victorges committed:
Organize imports

victorges committed:
Implement typing for the getIngest function

for that, implemented typing for the hardcoded-nodes
helper.

will continue by typing the kubernetes helper as well
for some level of completeness there.

victorges committed:
Make util file typescript

will need for kubernetes middleware

victorges committed:
Implement typing for Kubernetes middleware

victorges committed:
Further typing on stream controller

victorges committed:
Fix mist-api helper recursion call

I think there was a bug there using a previous
name of the function for the recursion :grimacing:

Viva TypeScript :tada:

victorges committed:
Some final touches on stream controller

noImplicitAny: true :sparkles:

victorges committed:
Add type-checking to the api package build pipeline

victorges committed:
Add missing TODOs

victorges committed:
@types/node-fetch global type hack

victorges committed:
Type parseCli and makeApp functions for real

This includes converting properties in returned object
from kebab to camel case :sparkles:

This makes the TestServer fully typed for stream tests
(yeah I went down that rabbit hole)

victorges committed:
Type all the stream tests :sparkles:

victorges committed:
Type session controller (is that still used?)

victorges committed:
Update ALL the typescript tooling

Seems like a good opportunity

(babel/core was actually necessary for some
of the new features being used, and typescript
is necessary for calling tsc)

victorges commented on packages/api/src/parse-cli.ts:
```diff
@@ -0,0 +1,220 @@
+import yargs from "yargs";
+import { CamelKeys } from "./types/common";
+
+function coerceArr(arg) {
```
Ugh just noticed Github didn't catch the rename here. I believe it's because the `parsed` intermediate variable in the `parceCli` function (used for typing purposes) caused a too-big indentation change.

Reviewer, let me know if you want me to put this change on a separate commit or something so it's more reviewable.

victorges replied:
OK, you can check the real changes made to this file, together with the related changes on this commit: 

https://github.com/livepeer/livepeer-com/pull/476/commits/0297730f626b4f3b0221d65120dbff8656639696?w=1

Link already ignoring whitespace changes which are basically prettier itself 👌

victorges committed:
Make parse-cli file typescript

Making this on a separate commit is necessary for easier review.

victorges committed:
Fix docker build

was missing tsconfig in the image

iameli commented on packages/api/Dockerfile:
```diff
@@ -6,19 +6,26 @@ RUN npm install -g --force yarn
+
 ADD package.json .
-RUN yarn install
+RUN yarn install --frozen-lockfile
```
Hmm, I'm surprised this works absent the presence of a yarn.lock file.

victorges replied:
lol good catch. I'll add the yarn.lock here too, also expected it would have given me an error there.

victorges replied:
So I fixed this by running the `docker build` with the context of the repository root instead, so I could copy the `yarn.lock` from the root `package.json`. Felt a bit awkward to be honest, wish there was an easier way, but I think it works fine. Let me know if you have a better idea or rather keep it the way it was, without `yarn.lock` (seems risky tho). I didn't really need this change to get the `tsconfig.json` to work, so I can revert the other changes I included.

Here's the commit: https://github.com/livepeer/livepeer-com/pull/476/commits/52dd0aff7f57c0bebdf231597ab700343cc9ef4b

iameli commented on packages/api/src/controllers/session.ts:
```diff
@@ -40,9 +45,25 @@ const fieldsMap = {
 app.get("/", authMiddleware({}), async (req, res, next) => {
-  let { limit, cursor, all, order, filters, userId, parentId, count } =
-    req.query;
+  let {
```
This just for typescript purposes? What would these things come in as if not for strings?

victorges replied:
This is actually not only a type issue, but the typing exposes a possible issue in code that only uses a `req.query.param` assuming it is always a `string`. The newer `@types/express` have a proper type for the `req.query` property which exposes the issue:
```
interface ParsedQs { [key: string]: undefined | string | string[] | ParsedQs | ParsedQs[] }
```

Which basically means that a `req.query` can contain a string, an array of strings or even recursive `ParsedQs` objects or array of those objects. That can happen on these cases:
 - `?cursor=abc&cursor=cde`
 - `?cursor[]=abc`
 - `?cursor[k1][k2]=value`
 - `?cursor[][k1][k2]=value` (?)

Long explanation here: https://evanhahn.com/gotchas-with-express-query-parsing-and-how-to-avoid-them/

Since I've found that issue, I've been using the `toString()` approach that I found somewhere, which seems the easier/safer to me. It doesn't explicitly give support to all possible corner cases that one might receive, but we probably don't want to give support to those anyway. It will work as expected if 1 value is sent or if an array is sent with only 1 element. In other cases it will generate a string that will cause an error somewhere else in the code (probably a 400, 422 or 404), but at least it won't explode a 500 due to possibly mistreating the object types. 

Anyway, if you feel like we should better handle all the `ParsedQs` possibilities let me know and I can improve that helper, maybe put it into a middleware which populates the `req` or at least in a common helper, idk. I didn't include that in this PR already as it was a bit massive already with the type changes.

iameli commented on packages/api/src/middleware/hardcoded-nodes.ts:
```diff
@@ -22,17 +39,17 @@ export default function hardcodedNodes({
     }
-    if (!req.prices) {
-      req.getPrices = async () => prices;
+    if (!req.getPrices) {
```
Hm, this check must have never worked with the wrong name there. I also don't really know what it's for. @yondonfu @kyriediculous?

victorges replied:
I think it really has never worked indeed, always overwriting what was in the request. I think it wasn't really an issue though because no other middleware overwrites that property. References in repo: https://github.com/livepeer/livepeer-com/search?q=getPrices

The `kubernetes` one actually [only sets `getBroadcasters` and `getOrchestrators`](https://github.com/livepeer/livepeer-com/blob/bd7528e04a91310fe2baa833d4504df0f96dde70/packages/api/src/middleware/kubernetes.ts#L107) so I'm guessing we only use the "hardcoded" ones configured on `LP_PRICES` env [in livepeer-infra](https://github.com/livepeer/livepeer-infra/blob/f78b771069043f14bc6fa79145cfdb392aee46fd/values/prod/livepeer.values.yaml#L5-L10)

iameli commented on packages/api/src/test-server.ts:
```diff
@@ -1,14 +1,15 @@
 import path from "path";
 import os from "os";
-import fetch from "isomorphic-fetch";
+import "isomorphic-fetch";
```
I know this works but dumping stuff on the global scope is a little frightening. Want to just use `node-fetch` instead?

victorges replied:
Switched to `node-fetch`!

Answering the comment though: It actually doesn't make a difference which way we import the library, as it always messes with the global scope when imported. So I had switched to not use the default export as I thought it was a bit misleading, using the value it sets on the global scope instead. I guess this import worked like a "blank identifier import" in go.

In fact:
 - `global.fetch` and `require("isomorphic-fetch")` are the exact same value as they do [`global.fetch = module.exports;` here](https://github.com/matthew-andrews/isomorphic-fetch/blob/master/fetch-npm-node.js#L12) (although we could argue that someone else might change the `global.fetch` after the lib has run)
 - Also learned that `isomorphic-fetch` basically exposes `node-fetch` on the global scope, doing like 1 `//`=>`https://` conversion before literally proxying to `node-fetch`. So the type hack I made doesn't even feel that much like a hack anymore 😆 (but it's much better to just use `node-fetch` directly anyway)

victorges replied:
Btw, the whole typing problem with `isomorphic-fetch` is that [their DefinitelyTyped package](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/isomorphic-fetch/index.d.ts#L8) assumes that it's going to run on a browser and have the `fetch` type in the global scope. So we then also had to transpile our node app with the `dom` module, which could lead to programming errors if you use something from the browser not available in node. Kinda non-sense if you ask me, given the purpose of this isomorphic package, but I'm also not a TS expert to know how to fix it easily on their repository. So let's just stick with `node-fetch` for now 😬

iameli reviewed (approved):
Sweet. Only actual comment is about the awkwardness around `isomorphic-fetch`.

victorges committed:
Import root project yarn.lock on docker build

This required actually changing the build context to
be the root directory and importing the relative paths to
the API package.

victorges committed:
Completely replace isomorphic-fetch with node-fetch

darkdarkdragon reviewed (approved):
LGTM

victorges committed:
Re-add createdAt field on stream.detection
